proposals/context-parameters.md

```diff
@@ -0,0 +1,560 @@
+
+**§14** *(contextual visibility)*: Concretely, whenever a context receiver is brought into scope, only the members with a declared context are available without qualification. Note that this includes those with an 0-ary context, which can be thought of as a "marker" for contextual visibility.
+
+```kotlin
+interface Logger {
+    context fun log(message: String)  // context() fun log(..)
+    fun thing()
+}
+
+fun Logger.shout(message: String) = log(message.uppercase())
+
+context fun Logger.whisper(message: String) = log(message.lowercase())
+
+context(Logger) fun logWithName(name: String, message: String) =
+    log("$name: $message")
+
+context(Logger) fun doSomething() {
+    log("Hello")  // ok
+    thing()  // unresolved reference 'thing'
+    summon<Logger>().thing()  // ok
+    shout("Hello")  // unresolved reference 'shout'
+    whisper("Hello")  // ok
+    logWithName("Alex", "Hello")  // ok
+}
+
+fun somethingElse() {
+    log("Hello")  // unresolved reference 'log'
+    logWithName("Alex", "Hello")  // no context receiver for 'Logger' found
+}
+```
+
+### The migration story for receivers
+
+**§15** *(receiver migration, caveats)*: It is **not** desirable to change existing APIs from extension receivers to context parameters, because of the potential changes in resolution (different meaning of `this`, changes in callable references). This is especially relevant for libraries which are consumed by virtually everybody (standard library, coroutines).
+
+**§16** *(receiver migration, scope types)*: If your library exposes a "scope" or "context" type, we suggest adding an empty context to their members as soon as possible. Every extension function over this type also needs that modifier.
+
+*How do you know that your type `S` is of such kind?*
+
+* It's often used as extension receiver, but without being the "subject" of the action.
+* There's some "run" function that takes a `S.() → R` function as an argument.
+
+
+The benefit for the consumer of the library is that now they can write functions where the type is a context receiver. This can lead to nicer-to-read code, since it "frees the extension position".
+
+```kotlin
+context(ApplicationContext) fun User.buildHtml() = ...
+```
+
+**§17** *(receiver migration, functions with subject)*: We encourage library authors to review their API, look for those cases, and provide new variants based on context parameters, instead of merely marking the original function with context.
+
+```kotlin
+// Uses extension receivers, but File is in fact the "subject"
+fun ResourceScope.openFile(file: File): InputStream
+
+// Doing this is OK, but provides no additonal benefit
+context fun ResourceScope.openFile(file: File): InputStream
+
+// The API is nicer using context receivers
+@JvmName("contextualFileOpen")  // needed to avoid platform clash
```

Comment:
```suggestion
```
why would there be a platform clash? the "old" function is named `openFile`, the "new" function is called `open`